### PR TITLE
Common tasks

### PR DESCRIPTION
**Pull request by cadenceforney**
closes issue #7

### CODE REVIEW

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/DemoAPI.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    // The wheel motors
+    // `lateinit` means the variable does not have to be initialized in the constructor
+    // See: https://github.com/BotsBurgh/BOTSBURGH-FTC-2023-24/pull/7#discussion_r1342154226
+    private lateinit var fl: DcMotor
+    private lateinit var fr: DcMotor
+    private lateinit var bl: DcMotor
+    private lateinit var br: DcMotor
+
+    // `init()` is run once at the beginning of the program.
+    // It initializes things and sets up configuration
```
Comment (BD103) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/DemoAPI.kt]:
```suggestion
    // `init()` is called by the opmode in order to use this API.
    // Override it if you need do things at startup, such as accessing hardware.
```

I would rephrase this to show that opmodes have to manually call `MyAPI.init()`. I feel like your original wording makes it seem like this is done automatically.

I wrote a suggestion that you can use or rephrase if you want!

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/DemoLinearAPI.kt [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import kotlin.math.atan2
+import kotlin.math.sqrt
+
+/**
+ * This is a demo API for driving related to april tags.
+ *
+ * Requires the [TriWheels] API.
+ */
+object DemoLinearAPI : API() {
+    override val isLinear = true
```
Comment (BD103) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/DemoLinearAPI.kt]:
```suggestion
    // `isLinear` requires that this API is only initialized with a `LinearOpMode`.
    // By enabling it, you can access `this.linearOpMode`.
    override val isLinear = true
```

`isLinear` is the largest thing that differentiates linear APIs from normal ones. I think it deserves a comment explaining what it does.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/DemoLinearAPI.kt [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import com.qualcomm.robotcore.util.Range
+import org.firstinspires.ftc.teamcode.api.TriWheels
+import org.firstinspires.ftc.teamcode.core.API
+import kotlin.math.atan2
+import kotlin.math.sqrt
+
+/**
+ * This is a demo API for driving related to april tags.
+ *
+ * Requires the [TriWheels] API.
```
Comment (BD103) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/DemoLinearAPI.kt]: APIs now have a dependency system so it will throw an exception if you initialize this and not `TriWheels`. Could you please override `this.dependencies` with a [set](https://kotlinlang.org/docs/collections-overview.html#set) containing `TriWheels`?

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/DemoLinearAPI.kt [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        // Treat range and strafe as an XY vector
+        // Then convert to the polar coordinate system (angle, length)
+        // This is the same math used to
+        val radians = atan2(range, strafe)
+        val magnitude = sqrt(range * range + strafe * strafe)
+
+        // Set the power of the wheels
+        TriWheels.drive(radians, magnitude, rotation = heading)
+    }
+    // Stop the wheels from moving, the target has been reached!
```
Comment (BD103) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/DemoLinearAPI.kt]: Should there be a `TriWheels.stop()` here?

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/autonomous/DemoAuto.kt
Comment (BD103) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/autonomous/DemoAuto.kt:1]: Could you move this to be within `teamcode/autonomous/demos`? It's what I did with [`EncoderLocalizationDemo`](https://github.com/BotsBurgh/BOTSBURGH-FTC-2024-25/blob/5b7de3bcf2c25bb54fa10700c4805a75807dc1e8/TeamCode/src/main/java/org/firstinspires/ftc/teamcode/teleop/demos/EncoderLocalizationDemo.kt), so that example code would not be confused with competition code.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/autonomous/DemoAuto.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
+package org.firstinspires.ftc.teamcode.opmode.autonomous
+
+import com.qualcomm.robotcore.eventloop.opmode.Autonomous
+import com.qualcomm.robotcore.eventloop.opmode.Disabled
+import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
+import com.qualcomm.robotcore.util.ElapsedTime
+import org.firstinspires.ftc.teamcode.api.DemoAPI
+import org.firstinspires.ftc.teamcode.api.linear.DemoLinearAPI
+
+@Autonomous(name = "Demo Auto")
```
Comment (BD103) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/autonomous/DemoAuto.kt]:
```suggestion
@Autonomous(name = "Demo Auto", group = "Demo")
```

I like to add a group to these opmodes so that they are all grouped together in the UI on the driver hub.

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/teleop/DemoTeleOp.kt
Comment (BD103) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/teleop/DemoTeleOp.kt:1]: Same thing as above, could you please move this within `teamcode/teleop/demos`?

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/teleop/DemoTeleOp.kt [hunk 6]
```diff
@@ -0,0 +1,35 @@
+package org.firstinspires.ftc.teamcode.opmode.teleop
+
+import com.qualcomm.robotcore.eventloop.opmode.Disabled
+import com.qualcomm.robotcore.eventloop.opmode.OpMode
+import com.qualcomm.robotcore.eventloop.opmode.TeleOp
+import org.firstinspires.ftc.teamcode.api.DemoAPI
+import org.firstinspires.ftc.teamcode.api.linear.DemoLinearAPI
+
+@TeleOp(name = "Demo TeleOp")
```
Comment (BD103) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/teleop/DemoTeleOp.kt]:
```suggestion
@TeleOp(name = "Demo TeleOp", group = "Demo")
```

Also same thing as above :)

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/teleop/DemoTeleOp.kt [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+@Disabled
+class DemoTeleOp : OpMode() {
+    // Run once, when "Init" is pressed on driver hub.
+    override fun init() {
+        // Initializes an API, shared code used by components.
+        DemoAPI.init(this)
+
+        // Initialize the April Vision API
+        DemoLinearAPI.init(this)
```
Comment (BD103) [inline TeamCode/src/main/java/org/firstinspires/ftc/teamcode/teleop/DemoTeleOp.kt]: This line will throw an exception because linear APIs cannot be initialized with `OpMode`s, only `LinearOpMode`s.

Could you either remove it or comment this line out, explaining why it won't work?

**Review by BD103 — CHANGES_REQUESTED** [review]
Really good start! I like how you use real examples that could actually be used, as compared to something more abstract. I have a few requests, but the majority of the work is done!

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/DemoLinearAPI.kt (hunk 3 above)
Comment (cadenceforney) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/DemoLinearAPI.kt]: I can not figure out how to write this line correctly.
I have tried different variations of:
override fun this.dependencies() = setOf(TriWheels)
override this.dependencies() = setOf(TriWheels)

Comment (BD103) [reply TeamCode/src/main/java/org/firstinspires/ftc/teamcode/api/DemoLinearAPI.kt]: This should work:

```kotlin
override val dependencies = setOf(TriWheels)
```

**Review by BD103 — APPROVED** [review]
Thank you! :D
